flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java

```diff
@@ -2260,6 +2263,121 @@ public void testSavepointScheduledInUnalignedMode() throws Exception {
 		}
 	}
 
+	/**
+	 * Test that the checkpoint still behave correctly when the task checkpoint is triggered by the
+	 * master hooks and finished before the master checkpoint.
+	 */
```

Comment:
So this test is covering for both bug fixes?

Reply:
The test covers the checkpoint failure issue. The other case is sort of indirectly tested by other tests which ensures all the `OperatorCoordinators` have finished snapshots before moving on to the next step. We can add another test to explicitly test the ordering issue.

Reply:
> The other case is sort of indirectly tested by other tests which ensures all the OperatorCoordinators have finished snapshots before moving on to the next step. 

But as you haven't modified those tests, and the master build was green, it seems like those tests were passing with the incorrect code.

Doesn't it mean we are still missing test coverage?

If so, that would mean splitting this change into to two separate bug fixes, with two separate test fixes would be much easier to understand and follow, which unit test is covering for what bug.

Reply:
> I end up with testing in one case for simplicity

It looks like the first bug fix from
```
final CompletableFuture<?> masterStatesComplete = pendingCheckpointCompletableFuture	
					.thenCompose(this::snapshotMasterState);
```
to
```
			final CompletableFuture<?> masterStatesComplete = coordinatorCheckpointsComplete
					.thenComposeAsync(ignored -> {
						PendingCheckpoint checkpoint =
							FutureUtils.getWithoutException(pendingCheckpointCompletableFuture);
						return snapshotMasterState(checkpoint);
					}, timer);
```
doesn't have a test coverage.

Could you split the two bug fixes into two commits? Besides that having both fixes in one commit is confusing it's also pretending that both are covered via the same unit test.

Reply:
I haven't got time to update the PR yet. The upcoming update has separated the PRs into two commits and added test coverage for both cases.